### PR TITLE
Tubs18 team2

### PR DESCRIPTION
**Pull request by BassMonster**
After a long time...

### CODE REVIEW

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * @author Sabrina Hugo
+ * @author Christian Orsinger
+ */
+public class CreateSiblingAction extends SingleSelectionAction {
+
+	public static final String ID = "de.ovgu.featureide.createsibling";
+
+	private final IGraphicalFeatureModel featureModel;
+
+	private static ImageDescriptor createImage = PlatformUI.getWorkbench().getSharedImages().getImageDescriptor(ISharedImages.IMG_OBJ_ADD);
```
Comment (KamilRosiak) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java:49]: Look in the GUIDefaults for PLUS_IMAGE.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	public CreateSiblingAction(Object viewer, IGraphicalFeatureModel featureModel) {
+		super(CREATE_SIBLING, viewer, ID);
+		setImageDescriptor(createImage);
+		this.featureModel = featureModel;
```
Comment (KamilRosiak) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java:54]: no null check for the featureModel

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		}
+	}
+
+	/**
+	 * @param selection selected elements
+	 * @return true, if only one element of FeatureEditPart or IFeature is selected and this is not the root
+	 */
+	@Override
+	protected boolean isValidSelection(IStructuredSelection selection) {
+		if (!super.isValidSelection(selection)) {
```
Comment (KamilRosiak) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java]: Check selection for null and for element in selection. (selection.isEmpty())

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+			return false;
+		}
+		final Object selectedObject = selection.getFirstElement();
+		IFeature selectedFeature;
+		if ((selectedObject instanceof FeatureEditPart) || (selectedObject instanceof IFeature)) {
+			if (selectedObject instanceof FeatureEditPart) {
```
Comment (KamilRosiak) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java]: duplicated code (same instanceof check above).

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	private final IGraphicalFeatureModel featureModel;
+	private IFeature newCompound;
+	private final int xDistanceTopDown = 5;
+	private final int yDistanceLeftRight = 8;
+	final IFeatureStructure parent;
+	private final int index;
+
+	public CreateSiblingOperation(IGraphicalFeatureModel featureModel, IFeature selectedFeature) {
+		super(featureModel.getFeatureModel(), CREATE_SIBLING);
+		this.featureModel = featureModel;
```
Comment (KamilRosiak) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java:56]: no null check

**Review by KamilRosiak — COMMENTED** [review]
Aside from some missing nullchecks it looks fine.
@PaulWestphal your turn.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java (hunk 1 above)
Comment (BassMonster) [reply plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateSiblingAction.java:49]: The PLUS_IMAGE in the GUIDefaults is of the type Image, but the needed type is ImageDescriptor, a cast isn't possible.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java [hunk 6]
```diff
@@ -169,8 +171,10 @@
 	private Label infoLabel;
 
 	private CalculateDependencyAction calculateDependencyAction;
-	private CreateLayerAction createLayerAction;
-	private CreateCompoundAction createCompoundAction;
+	private CreateFeatureBelowAction createLayerAction;
+	private CreateFeatureAboveAction createCompoundAction;
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java]: Change name of createLayerAction and createCompoundAction. Make sure to change the name everywhere it's used.

File: plugins/de.ovgu.featureide.Cloneanalysis/.project [hunk 7]
```diff
@@ -1,28 +1,28 @@
...
+			<name>org.eclipse.pde.SchemaBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+	</natures>
+</projectDescription>
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.Cloneanalysis/.project:28]: This shouldn't be in this pull request.

File: plugins/de.ovgu.featureide.Cloneanalysis/.settings/org.eclipse.jdt.core.prefs [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
+org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.6
+org.eclipse.jdt.core.compiler.codegen.unusedLocal=preserve
+org.eclipse.jdt.core.compiler.compliance=1.6
+org.eclipse.jdt.core.compiler.debug.lineNumber=generate
+org.eclipse.jdt.core.compiler.debug.localVariable=generate
+org.eclipse.jdt.core.compiler.debug.sourceFile=generate
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.source=1.6
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.Cloneanalysis/.settings/org.eclipse.jdt.core.prefs:11]: This shouldn't be in this pull request.

File: plugins/de.ovgu.featureide.Cloneanalysis/build.properties [hunk 9]
```diff
@@ -1,33 +1,33 @@
...
+               lib/pmd-ruby-5.2.2.jar,\
+               lib/pmd-vm-5.2.2.jar,\
+               lib/pmd-xml-5.2.2.jar,\
+               lib/rhino-1.7R4.jar,\
+               lib/saxon-9.1.0.8-dom.jar,\
+               lib/saxon-9.1.0.8.jar,\
+               lib/xercesImpl-2.11.0.jar,\
+               lib/xml-apis-1.4.01.jar,\
+               lib/jcommon-1.0.23.jar,\
+               lib/jfreechart-1.0.19.jar
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.Cloneanalysis/build.properties:33]: This shouldn't be in this pull request.

File: plugins/de.ovgu.featureide.Cloneanalysis/plugin.xml [hunk 10]
```diff
@@ -1,115 +1,115 @@
...
+  point="org.eclipse.ui.ide.markerResolution">
+    <markerResolutionGenerator
+        class="de.ovgu.featureide.cloneanalysis.utils.QuickFixer"
+        markerType="org.eclipse.core.resources.problemmarker">
+        
+</markerResolutionGenerator>
+</extension>
+
+
+</plugin>
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.Cloneanalysis/plugin.xml:115]: This shouldn't be in this pull request.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureBelowAction.java [hunk 11]
```diff
@@ -37,23 +38,26 @@
  *
  * @author Thomas Thuem
  */
-public class CreateLayerAction extends SingleSelectionAction {
+public class CreateFeatureBelowAction extends SingleSelectionAction {
 
 	public static final String ID = "de.ovgu.featureide.createlayer";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureBelowAction.java]: Change the ID.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureBelowAction.java [hunk 12]
```diff
@@ -37,23 +38,26 @@
...
 
 	private static ImageDescriptor createImage = PlatformUI.getWorkbench().getSharedImages().getImageDescriptor(ISharedImages.IMG_OBJ_ADD);
 
-	private final IGraphicalFeatureModel featureModel;
+	private final IFeatureModel featureModel;
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureBelowAction.java]: This is not needed anymore.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+		int number = 0;
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java]: Choose a more descriptive variable name.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+		while (FeatureUtils.getFeatureNames(featureModel.getFeatureModel()).contains(DEFAULT_FEATURE_LAYER_CAPTION + ++number)) {}
+
+		newCompound =
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/operations/CreateSiblingOperation.java]: We wanted to move away from calling things "compounds" and "layers".

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 15]
```diff
@@ -505,6 +505,7 @@
 	public static final String IT_CAN_NOT_BE_REPLACED_WITH_AN_EQUIVALENT_ONE_ = "It can not be replaced with an equivalent one.";
 	public static final String CHANGE_GROUP_TYPE = "Change Group Type";
 	public static final String CREATE_COMPOUND = "Create Compound";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java]: You should change this name while you're at it.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java [hunk 16]
```diff
@@ -608,6 +613,20 @@ public void propertyChange(FeatureIDEEvent event) {
 			setDirty();
 			analyzeFeatureModel();
 			break;
+		case FEATURE_ADD_SIBLING:
+			if ((event.getNewValue() != null) && (event.getNewValue() instanceof IFeature)) {
+				newCompound = (IFeature) event.getNewValue();
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java]: Don't call it compound.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureAboveAction.java [hunk 17]
```diff
@@ -51,12 +50,10 @@
  * @author Thomas Thuem
  * @author Marcus Pinnecke (Feature Interface)
  */
-public class CreateCompoundAction extends Action {
+public class CreateFeatureAboveAction extends Action {
 
 	public static final String ID = "de.ovgu.featureide.createcompound";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateFeatureAboveAction.java]: You should also change this ID.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 18]
```diff
@@ -505,6 +505,7 @@
...
+	public static final String CREATE_SIBLING = "Create Sibling Feature";
 	public static final String CREATE_LAYER = "Create Layer";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java]: Another outdated name.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java [hunk 19]
```diff
@@ -252,10 +256,11 @@ public FeatureDiagramEditor(IFileManager<IFeatureModel> fmManager, IFileManager<
 	private void createActions() {
 		final IFeatureModel featureModel = getFeatureModel();
 		actions.clear();
-
+		createLayerAction = addAction(new CreateFeatureBelowAction(viewer, graphicalFeatureModel));
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java]: Outdated name.

**Review by PaulWestphal — CHANGES_REQUESTED** [review]
You renamed the compound and layer actions, so you should apply these changes everywhere consistently.

**Comment by BassMonster** [comment]
@PaulWestphal would you please be so kind to merge this pull request, so that we can request the new changes

**Pull request merged**